### DISCUSSION TITLE
Simulating an ARM multicore system with more than 8 cores

### DISCUSSION THREAD

**Discussion by mbabaie** [gem5/gem5#1757] [Q&A]
Hello @giactra,

I'm reaching out regarding an issue I recently faced with when trying to simulate a large multicore system (e.g., hundreds of cores if possible) with ARM ISA using gem5's standard library components. I realized that some of the available infrastructures only supports up-to 8 cores. For instance, the platforms available (i.e., VExpress_GEM5_V1 or V2, etc) they only support up-to 8 cores. Do you know any solutions on how to possibly scale up the system to more than 8 cores?

Best regards,
Maryam

**Comment by giactra** [comment]
Hi @mbabaie .

This has partly been addressed by https://github.com/orgs/gem5/discussions/821 (for sure there are other references in the gem5 mailing list).
TL;DR: it is definitely possible to have > 8 cores with VExpress_GEM5_V2 and VExpress_GEM5_Foundation. This you can do as of now with a vanilla Linux kernel and no changes needed.

If you want to use VExpress_GEM5_V1, you need to run a modified GIC Linux driver to allow more than 8 cores.

**Reply by mbabaie** [reply]
Thanks for your prompt reply.

I tried the `VExpress_GEM5_Foundation` with 8-core and 16-core systems with KVM CPU. Is it normal that for a long time nothing is printed on the board.terminal? I don't know how long l should let it run. My assumption was that perhaps it's not booting linux.

**Reply by giactra** [reply]
Just to be clear: is using KVM your ultimate goal or are you trying it just because it is referenced in the discussion I posted?

**Reply by mbabaie** [reply]
I aim to use KVM to accelerate booting linux and take a checkpoint.
I have tested the `VExpress_GEM5_Foundation` with 16-cores with `atomic` CPU and it successfully booted and took the checkpoint. However, it's very slow and took about 12 hours to finish the simulation. KVM will make it much faster.

**Comment by powerjg** [comment]
@giactra here's the crux of my question: Can you use Arm+KVM with more than 8 cores? If so, how?

My hardware is an Ampere Altra. As far as I can tell, it is a Neoverse N1 with GICv3.

(Edit) I'm using gem5 version 24.1.0.1 (current stable). (I can't get develop to build because of #1923)

I have modified this line in `arm-ubuntu-run-with-kvm.py` to use `VExpress_GEM5_Foundation` instead of `VExpress_GEM5_V1`. https://github.com/gem5/gem5/blob/stable/configs/example/gem5_library/arm-ubuntu-run-with-kvm.py#L97

Things don't *fail*, but nothing is ever printed to the terminal and the readfile command (`m5 exit`) is never executed. It seems to be stuck spinning *before* the Linux boot process.

Any pointers to how to get KVM + more than 8 cores working would be greatly appreciated. I've attached the output of running the script linked above with the specified change.
[out.txt](https://github.com/user-attachments/files/18593005/out.txt)

**Reply by powerjg** [reply]
@giactra any information about this would be greatly appreciated!

**Reply by giactra** [reply]
Sorry @powerjg I have been quite busy. Will have a look at this sometime next week

**Reply by powerjg** [reply]
@giactra I meant to follow up on this a while ago. Here's the most detailed information I can give.

When I run the following things work: (Note, this is a slightly modified version of `configs/example/gem5_library/arm-ubuntu-run-with-kvm.py`)

```python
from m5.objects import (
    ArmDefaultRelease,
    Armv84,
    VExpress_GEM5_Foundation,
    VExpress_GEM5_V1,
)

from gem5.components.boards.arm_board import ArmBoard
from gem5.components.memory import DualChannelDDR4_2400
from gem5.components.processors.cpu_types import CPUTypes
from gem5.components.processors.simple_switchable_processor import (
    SimpleSwitchableProcessor,
)
from gem5.isas import ISA
from gem5.resources.resource import obtain_resource
from gem5.simulate.exit_event import ExitEvent
from gem5.simulate.simulator import Simulator

from gem5.components.cachehierarchies.classic.private_l1_private_l2_cache_hierarchy import (
    PrivateL1PrivateL2CacheHierarchy,
)

# Here we setup the board. The ArmBoard allows for Full-System ARM simulations.
board = ArmBoard(
    clk_freq="3GHz",
    processor=SimpleSwitchableProcessor(
        starting_core_type=CPUTypes.KVM,
        switch_core_type=CPUTypes.TIMING,
        isa=ISA.ARM,
        num_cores=2,
    ),
    memory=DualChannelDDR4_2400(size="2GiB"),
    cache_hierarchy=PrivateL1PrivateL2CacheHierarchy(
        l1d_size="16KiB", l1i_size="16KiB", l2_size="256KiB"
    ),
    release=ArmDefaultRelease.for_kvm(),
    platform=VExpress_GEM5_V1(),
)
board.set_workload(obtain_resource("arm-ubuntu-24.04-boot-with-systemd"))


def exit_event_handler():
    print("First exit: kernel booted")
    yield False  # gem5 is now executing systemd startup
    print("Second exit: Started `after_boot.sh` script")
    print("Switching to Timing CPU")
    board.processor.switch()
    yield False  # gem5 is now executing the `after_boot.sh` script
    print("Third exit: Finished `after_boot.sh` script")
    yield True


simulator = Simulator(
    board=board,
    on_exit_event={
        # Here we want override the default behavior for the first m5 exit
        # exit event.
        ExitEvent.EXIT: exit_event_handler()
    },
)

simulator.run()
```

When I change the platform to `VExpress_GEM5_Foundation` nothing happens. The simulation starts, but nothing is ever printed to any terminal devices. I do get the following warnings, though.

```
src/dev/arm/rv_ctrl.cc:176: warn: SCReg: Access to unknown device dcc0:site0:pos0:fn7:dev0
src/dev/arm/gic_v3_distributor.cc:860: warn: Gicv3Distributor::write(): setting ARE to 0 is not supported!
```

The following script doesn't work.

```python
from m5.objects import (
    ArmDefaultRelease,
    Armv84,
    VExpress_GEM5_Foundation,
    VExpress_GEM5_V1,
)

from gem5.components.boards.arm_board import ArmBoard
from gem5.components.memory import DualChannelDDR4_2400
from gem5.components.processors.cpu_types import CPUTypes
from gem5.components.processors.simple_switchable_processor import (
    SimpleSwitchableProcessor,
)
from gem5.isas import ISA
from gem5.resources.resource import obtain_resource
from gem5.simulate.exit_event import ExitEvent
from gem5.simulate.simulator import Simulator

from gem5.components.cachehierarchies.classic.private_l1_private_l2_cache_hierarchy import (
    PrivateL1PrivateL2CacheHierarchy,
)

# Here we setup the board. The ArmBoard allows for Full-System ARM simulations.
board = ArmBoard(
    clk_freq="3GHz",
    processor=SimpleSwitchableProcessor(
        starting_core_type=CPUTypes.KVM,
        switch_core_type=CPUTypes.TIMING,
        isa=ISA.ARM,
        num_cores=2,
    ),
    memory=DualChannelDDR4_2400(size="2GiB"),
    cache_hierarchy=PrivateL1PrivateL2CacheHierarchy(
        l1d_size="16KiB", l1i_size="16KiB", l2_size="256KiB"
    ),
    release=ArmDefaultRelease.for_kvm(),
    platform=VExpress_GEM5_Foundation(),
)
board.set_workload(obtain_resource("arm-ubuntu-24.04-boot-with-systemd"))


def exit_event_handler():
    print("First exit: kernel booted")
    yield False  # gem5 is now executing systemd startup
    print("Second exit: Started `after_boot.sh` script")
    print("Switching to Timing CPU")
    board.processor.switch()
    yield False  # gem5 is now executing the `after_boot.sh` script
    print("Third exit: Finished `after_boot.sh` script")
    yield True


simulator = Simulator(
    board=board,
    on_exit_event={
        # Here we want override the default behavior for the first m5 exit
        # exit event.
        ExitEvent.EXIT: exit_event_handler()
    },
)

simulator.run()
```

I have played around with the "Release" objects, but that didn't help. I have tried a few other things as well. Any help would be greatly appreciated! I'm confident that there's just something a little wrong in the [`ArmBoard`](https://github.com/gem5/gem5/blob/stable/src/python/gem5/components/boards/arm_board.py), but I don't know what it may be.

**Reply by powerjg** [reply]
@giactra here's a polite reminder :)

**Comment by h1demasa** [comment]
Any progress?
